### ISSUE TITLE
[Papers] Implement template training infrastructure

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ProjectOdyssey#2733]
## Summary
Implement paper template training components.

## Components Needed
- Argument parsing (when Mojo stdlib supports it)
- Model creation from config
- Trainer creation from config
- Training loop execution

## Files Affected
- \`papers/_template/examples/train.mojo\` (lines 128-131)

## Related TODO Lines in todo.log
- Lines 128-131: template training TODOs

## Notes
Each paper implementation will customize these patterns.

**Comment by mvillmow** [comment]
## Detailed Implementation Plan

Now I have a comprehensive understanding of the codebase. Let me create the detailed implementation plan:

---

# Implementation Plan: Issue #2733 - Template Training Infrastructure

## Summary

This issue addresses implementing the template training infrastructure for the ML Odyssey papers template. The goal is to replace four placeholder TODO implementations in `/home/mvillmow/ml-odyssey/papers/_template/examples/train.mojo` with functional code that demonstrates the configuration-driven training pattern.

The implementation will leverage existing infrastructure:
- **Argument parsing**: `shared/utils/arg_parser.mojo` and `shared/utils/training_args.mojo`
- **Configuration system**: `shared/utils/config.mojo` and `shared/utils/config_loader.mojo`
- **Trainer infrastructure**: `shared/training/trainer_interface.mojo`
- **Training utilities**: Optimizers, schedulers, metrics in `shared/training/`

The template serves as a reference implementation that each paper (LeNet-5, AlexNet, VGG-16, etc.) will customize.

---

## Step-by-Step Implementation Tasks

### Phase 1: Argument Parsing Implementation (TODO #1)

**File**: `/home/mvillmow/ml-odyssey/papers/_template/examples/train.mojo` (lines 22-34)

**Task 1.1**: Extend the argument parser for template-specific arguments

1. Import `create_training_parser` from `shared.utils.arg_parser`
2. Add template-specific arguments:
   - `--paper <string>`: Paper name (e.g., "lenet5", "alexnet") - default: "lenet5"
   - `--experiment <string>`: Experiment name (e.g., "baseline", "augmented") - default: "baseline"
3. Parse arguments and return tuple `(paper_name, experiment_name)`

**Implementation Pattern** (from AlexNet example):
```mojo
fn parse_arguments() raises -> (String, String):
    """Parse command-line arguments for paper and experiment selection.
    
    Returns:
        Tuple of (paper_name, experiment_name)
    """
    var parser = create_training_parser()
    parser.add_argument("paper", "string", "lenet5")
    parser.add_argument("experiment", "string", "baseline")
    
    var args = parser.parse()
    var paper_name = args.get_string("paper", "lenet5")
    var experiment_name = args.get_string("experiment", "baseline")
    
    return (paper_name, experiment_name)
```

### Phase 2: Model Creation Implementation (TODO #2)

**File**: `/home/mvillmow/ml-odyssey/papers/_template/examples/train.mojo` (lines 37-65)

**Task 2.1**: Implement model factory function

1. Extract model configuration from Config object:
   - `model.name`: Model type identifier
   - `model.num_classes`: Number of output classes
   - `model.input_shape`: Input tensor shape
   - `model.dtype`: Data type (default: float32)
   - `model.dropout`: Dropout rate

2. Create a `ModelConfig` struct to hold model parameters

3. Implement model factory pattern that:
   - Validates required config keys exist
   - Returns appropriate model based on `model.name`
   - Handles errors gracefully

**Implementation Pattern**:
```mojo
struct ModelConfig(Copyable, Movable):
    """Configuration for model instantiation."""
    var name: String
    var num_classes: Int
    var input_shape: List[Int]
    var dtype: DType
    var dropout: Float32

fn create_model(config: Config) raises -> ModelConfig:
    """Create model configuration from config file.
    
    Args:
        config: Configuration containing model parameters
        
    Returns:
        ModelConfig with validated parameters
        
    Raises:
        Error if required configuration is missing
    """
    # Validate required keys
    if not config.has("model.name"):
        raise Error("Missing required config: model.name")
    
    var model_config = ModelConfig(
        name=config.get_string("model.name"),
        num_classes=config.get_int("model.num_classes", default=10),
        input_shape=parse_input_shape(config),
        dtype=DType.float32,  # Default dtype
        dropout=Float32(config.get_float("model.dropout", default=0.0))
    )
    
    print("Creating model from configuration...")
    print("  Model name:", model_config.name)
    print("  Number of classes:", model_config.num_classes)
    print("  Input shape:", format_shape(model_config.input_shape))
    
    return model_config
```

### Phase 3: Trainer Creation Implementation (TODO #3)

**File**: `/home/mvillmow/ml-odyssey/papers/_template/examples/train.mojo` (lines 67-90)

**Task 3.1**: Implement trainer factory function

1. Extract training configuration:
   - `optimizer.name`: Optimizer type (sgd, adam, adamw)
   - `optimizer.learning_rate`: Learning rate
   - `optimizer.momentum`: Momentum (for SGD)
   - `optimizer.weight_decay`: Weight decay
   - `training.epochs`: Number of epochs
   - `training.batch_size`: Batch size
   - `logging.log_interval`: Log interval

2. Create `TrainerConfig` from `shared.training.trainer_interface`

3. Return configured trainer

**Implementation Pattern**:
```mojo
from shared.training.trainer_interface import TrainerConfig

fn create_trainer(config: Config) raises -> TrainerConfig:
    """Create trainer configuration from config file.
    
    Args:
        config: Configuration containing training parameters
        
    Returns:
        TrainerConfig with validated parameters
    """
    var learning_rate = config.get_float("optimizer.learning_rate", default=0.001)
    var batch_size = config.get_int("training.batch_size", default=32)
    var epochs = config.get_int("training.epochs", default=10)
    var log_interval = config.get_int("logging.log_interval", default=10)
    var save_checkpoints = config.get_bool("logging.save_checkpoints", default=False)
    var checkpoint_interval = config.get_int("logging.checkpoint_interval", default=5)
    
    var trainer_config = TrainerConfig(
        num_epochs=epochs,
        batch_size=batch_size,
        learning_rate=learning_rate,
        log_interval=log_interval,
        save_checkpoints=save_checkpoints,
        checkpoint_interval=checkpoint_interval
    )
    
    var optimizer_name = config.get_string("optimizer.name", default="sgd")
    print("Creating trainer from configuration...")
    print("  Optimizer:", optimizer_name)
    print("  Learning rate:", learning_rate)
    print("  Batch size:", batch_size)
    print("  Epochs:", epochs)
    
    return trainer_config
```

### Phase 4: Training Loop Implementation (TODO #4)

**File**: `/home/mvillmow/ml-odyssey/papers/_template/examples/train.mojo` (lines 92-112)

**Task 4.1**: Implement training loop skeleton

1. Use `TrainingMetrics` from trainer interface for tracking
2. Implement epoch loop with:
   - Batch iteration (placeholder until DataLoader is integrated)
   - Loss tracking
   - Validation at intervals
   - Checkpoint saving

**Implementation Pattern**:
```mojo
from shared.training.trainer_interface import TrainingMetrics

fn run_training(
    config: Config,
    model_config: ModelConfig,
    trainer_config: TrainerConfig
) raises:
    """Run the training loop.
    
    Args:
        config: Complete configuration
        model_config: Model configuration
        trainer_config: Trainer configuration
    """
    print("\nStarting training...")
    
    var metrics = TrainingMetrics()
    var epochs = trainer_config.num_epochs
    var log_interval = trainer_config.log_interval
    var validate_interval = config.get_int("training.validate_interval", default=1)
    
    for epoch in range(epochs):
        metrics.current_epoch = epoch + 1
        metrics.reset_epoch()
        
        print("Epoch", epoch + 1, "/", epochs)
        
        # Training phase (placeholder - actual model forward/backward goes here)
        # In real implementation, iterate over DataLoader batches:
        # while train_loader.has_next():
        #     var batch = train_loader.next()
        #     var loss = compute_gradients(model, batch.data, batch.labels, ...)
        #     metrics.update_train_metrics(Float64(loss), 0.0)
        
        # Placeholder loss computation
        var train_loss = 1.0 / Float64(epoch + 1)  # Simulate decreasing loss
        metrics.update_train_metrics(train_loss, 0.0)
        
        print("  Train Loss:", train_loss)
        
        # Validation phase
        if validate_interval > 0 and (epoch + 1) % validate_interval == 0:
            # Placeholder validation - real impl would run on validation set
            var val_loss = train_loss * 1.1
            metrics.update_val_metrics(val_loss, 0.0)
            print("  Val Loss:", val_loss)
        
        # Checkpoint saving
        if trainer_config.save_checkpoints:
            if (epoch + 1) % trainer_config.checkpoint_interval == 0:
                print("  Saving checkpoint...")
                # save_model_weights(model_params, checkpoint_dir, param_names)
    
    print("\nTraining complete!")
    metrics.print_summary()
```

### Phase 5: Update Main Function

**Task 5.1**: Wire everything together in main()

1. Update main() to use new function signatures
2. Pass model_config and trainer_config between functions
3. Add error handling for each step

```mojo
fn main() raises:
    """Main entry point for training script."""
    print("=" * 60)
    print("Configuration-Driven Training Example")
    print("=" * 60)

    # Step 1: Parse arguments
    var args = parse_arguments()
    var paper_name = args[0]
    var experiment_name = args[1]

    print("\nPaper:", paper_name)
    print("Experiment:", experiment_name)

    # Step 2: Load configuration
    print("\nLoading configuration...")
    var config = load_experiment_config(paper_name, experiment_name)
    print("  Configuration loaded successfully!")

    # Step 3: Validate configuration
    print("\nValidating configuration...")
    try:
        validate_experiment_config(config)
        print("  Configuration is valid!")
    except e:
        print("  ERROR: Configuration validation failed!")
        print("  ", String(e))
        return

    # Step 4: Create model from config
    var model_config = create_model(config)

    # Step 5: Create trainer from config
    var trainer_config = create_trainer(config)

    # Step 6: Run training
    run_training(config, model_config, trainer_config)

    print("\n" + "=" * 60)
    print("Training script completed successfully!")
    print("=" * 60)
```

---

## Files to Modify/Create

### Files to Modify

| File | Changes | Lines Affected |
|------|---------|----------------|
| `/home/mvillmow/ml-odyssey/papers/_template/examples/train.mojo` | Implement all 4 TODOs | Lines 22-34, 37-65, 67-90, 92-112, 114-154 |

### Files to Create (Optional Enhancement)

| File | Purpose |
|------|---------|
| `/home/mvillmow/ml-odyssey/papers/_template/src/model_factory.mojo` | Model factory utilities (if factored out) |
| `/home/mvillmow/ml-odyssey/papers/_template/tests/test_train.mojo` | Unit tests for training components |

### Related Files (Reference Only)

| File | Purpose |
|------|---------|
| `/home/mvillmow/ml-odyssey/shared/utils/arg_parser.mojo` | Argument parsing infrastructure |
| `/home/mvillmow/ml-odyssey/shared/utils/training_args.mojo` | TrainingArgs struct |
| `/home/mvillmow/ml-odyssey/shared/utils/config.mojo` | Configuration system |
| `/home/mvillmow/ml-odyssey/shared/utils/config_loader.mojo` | Config loading utilities |
| `/home/mvillmow/ml-odyssey/shared/training/trainer_interface.mojo` | TrainerConfig, TrainingMetrics |
| `/home/mvillmow/ml-odyssey/configs/defaults/*.yaml` | Default configurations |
| `/home/mvillmow/ml-odyssey/papers/_template/configs/config.yaml` | Template config example |

---

## Testing Approach

### Unit Tests

**Test File**: `/home/mvillmow/ml-odyssey/papers/_template/tests/test_train.mojo`

1. **Test argument parsing**:
   - Default values returned when no args provided
   - Custom values parsed correctly from simulated args
   - Error handling for invalid arguments

2. **Test model configuration extraction**:
   - Required config keys validated
   - Default values used when optional keys missing
   - Error raised for missing required keys

3. **Test trainer configuration extraction**:
   - All TrainerConfig fields populated correctly
   - Default values applied appropriately
   - Mixed precision settings extracted if present

4. **Test training loop**:
   - Metrics updated correctly per epoch
   - Validation runs at correct intervals
   - Checkpoint logic triggered appropriately

### Integration Tests

1. **Test full pipeline**:
   - Run train.mojo with sample config
   - Verify configuration loads correctly
   - Verify training completes without error

2. **Test config cascade**:
   - Create paper-specific and experiment-specific configs
   - Verify merge order (defaults → paper → experiment)

### Manual Verification

```bash
# Test with default arguments
cd /home/mvillmow/ml-odyssey
pixi run mojo papers/_template/examples/train.mojo

# Test with custom arguments
pixi run mojo papers/_template/examples/train.mojo --paper lenet5 --experiment baseline

# Test with training args
pixi run mojo papers/_template/examples/train.mojo --epochs 5 --batch-size 64 --lr 0.01
```

---

## Success Criteria

### Functional Requirements

- [ ] **TODO #1 Resolved**: `parse_arguments()` successfully parses `--paper` and `--experiment` arguments using `create_training_parser()`
- [ ] **TODO #2 Resolved**: `create_model()` extracts model configuration and returns `ModelConfig` struct
- [ ] **TODO #3 Resolved**: `create_trainer()` extracts training configuration and returns `TrainerConfig` struct
- [ ] **TODO #4 Resolved**: `run_training()` implements epoch loop with metrics tracking and validation

### Quality Requirements

- [ ] Code follows Mojo v0.25.7+ syntax (out self, mut, etc.)
- [ ] Docstrings present for all functions
- [ ] Error handling for missing config keys
- [ ] No hardcoded values - all parameters from config
- [ ] Compiles without warnings: `pixi run mojo build papers/_template/examples/train.mojo`

### Template Requirements

- [ ] Serves as reference for paper implementations
- [ ] Pattern is generalizable (works with any paper/experiment)
- [ ] Comments explain customization points
- [ ] Integration with existing shared infrastructure

### CI/CD Requirements

- [ ] Pre-commit hooks pass: `pre-commit run --all-files`
- [ ] Mojo formatting passes: `pixi run mojo format papers/_template/examples/train.mojo`
- [ ] Tests pass (when created): `pixi run mojo test papers/_template/tests/`

---

## Dependencies

### Blocking Dependencies

1. **Configuration system** (✅ Complete):
   - `shared/utils/config.mojo` - Config struct with YAML/JSON support
   - `shared/utils/config_loader.mojo` - load_experiment_config, validate_experiment_config

2. **Argument parsing** (✅ Complete):
   - `shared/utils/arg_parser.mojo` - ArgumentParser, create_training_parser
   - `shared/utils/training_args.mojo` - TrainingArgs struct

3. **Trainer interface** (✅ Complete):
   - `shared/training/trainer_interface.mojo` - TrainerConfig, TrainingMetrics, DataLoader

### Non-Blocking Dependencies (Future Enhancement)

1. **Autograd** (⏳ In Progress):
   - Once available, training loop can use automatic gradient computation
   - Currently requires manual backward pass implementation per model

2. **Data loading** (✅ Available):
   - `shared/data/datasets.py` - Dataset loading utilities
   - DataLoader in trainer_interface for batch iteration

---

## Implementation Notes

### Mojo Stdlib Limitations

Per TODO #1, Mojo stdlib argument parsing is limited. The current implementation uses:
- Custom `ArgumentParser` in `shared/utils/arg_parser.mojo`
- String-based argument parsing with type conversion
- Works for current use case; stdlib support may improve in future

### Template Customization Points

Each paper implementation will customize:

1. **Model creation**: Import specific model (LeNet5, AlexNet, etc.)
2. **Backward pass**: Implement `compute_gradients()` with model-specific layers
3. **Data loading**: Use appropriate dataset loader
4. **Configuration**: Paper-specific config in `configs/papers/{paper_name}/`

### Design Decisions

1. **Return structs, not models**: Functions return configuration structs rather than instantiated models. Actual model instantiation happens in paper-specific code.

2. **Composition over inheritance**: Use TrainerConfig + functions rather than subclassing a base trainer.

3. **Explicit over implicit**: All parameters come from config files, making behavior transparent and reproducible.

---

## Resource Usage

| Metric | Count |
|--------|-------|
| Tool Calls | 12 |
| Files Read | 10 |
| Files Explored | ~30 (via agents) |
| Steps | 6 |
